### PR TITLE
WIP: Fuse-BSC Alm adaption

### PR DESCRIPTION
**Pull request by t0mcr8se**

### CODE REVIEW

File: alm/.env.example [hunk 1]
```diff
@@ -1,15 +1,24 @@
-COMMON_HOME_BRIDGE_ADDRESS=0x2CA5411c4bf447Cc27CD6E6d1d046f922A27C399
-COMMON_FOREIGN_BRIDGE_ADDRESS=0x63C47c296B63bE888e9af376bd927C835014039f
+COMMON_HOME_BRIDGE_ADDRESS=0x1ee6E3E3d2DE779858728E157B3B9C488bA7b706
```
Comment (leonprou) [inline alm/.env.example]: can you rename this file to `.env.bsc.example`?
and leave the original one untouched

File: alm/.env.example [hunk 2]
```diff
@@ -1,15 +1,24 @@
...
-ALM_FOREIGN_EXPLORER_TX_TEMPLATE=https://etherscan.io/tx/%s
+ALM_FOREIGN_EXPLORER_TX_TEMPLATE=https://bscscan.io/tx/%s
 
 ALM_HOME_EXPLORER_API=https://explorer.fuse.io/api
-ALM_FOREIGN_EXPLORER_API=https://etherscan.io/api?apikey=HFXDYDZ385WC8UZM6KJTRHUY5IBNAKX99J
+ALM_FOREIGN_EXPLORER_API=https://bscscan.io/api?apikey=DMVN9TSF99IVZRGYM6WYDJPGE7N72JCGG7
 PORT=8080
+
+# Block limit of RPC for getLogs method (-1 indicates no block limit)
+HOME_BRIDGE_RPC_BLOCK_LIMIT=-1
```
Comment (leonprou) [inline alm/.env.example]: can you use true/false values for this?

File: alm/src/utils/limitedEvents.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import { Contract } from 'web3-eth-contract'
+import { EventData } from 'web3-eth-contract'
+import { FOREIGN_BLOCK_RANGE_LIMIT, HOME_BLOCK_RANGE_LIMIT, HOME_RPC_URL, FOREIGN_RPC_URL } from '../config/constants'
+import { getWeb3 } from './web3'
+
+export const getPastEventsLimited = async (
```
Comment (leonprou) [inline alm/src/utils/limitedEvents.ts:6]: you got his implemented twice, no?

File: alm/scripts/createSnapshots.js [hunk 4]
```diff
@@ -26,13 +52,24 @@ const generateSnapshot = async (side, url, bridgeAddress) => {
...
+
   const bridgeContract = new web3.eth.Contract(HOME_AMB_ABI, bridgeAddress)
 
   // Save RequiredBlockConfirmationChanged events
-  let requiredBlockConfirmationChangedEvents = await bridgeContract.getPastEvents('RequiredBlockConfirmationChanged', {
-    fromBlock: 0,
-    toBlock: currentBlockNumber
-  })
+
+  // let requiredBlockConfirmationChangedEvents = await bridgeContract.getPastEvents('RequiredBlockConfirmationChanged', {
```
Comment (leonprou) [inline alm/scripts/createSnapshots.js]: delete the commented code

File: alm/scripts/createSnapshots.js [hunk 5]
```diff
@@ -5,15 +5,41 @@ require('dotenv').config()
...
+  HOME_BRIDGE_RPC_BLOCK_LIMIT,
+  FOREIGN_BRIDGE_DEPLOYED_AT_BLOCK
 } = process.env
 
-const generateSnapshot = async (side, url, bridgeAddress) => {
+const getPastEventsLimited = async (contract, eventName, fromBlock, toBlock, blockLimit) => {
+  let events = []
+  const chunks = parseInt(blockLimit)
+
+  if (chunks === -1) {
```
Comment (leonprou) [inline alm/scripts/createSnapshots.js]: it can be optional param, with undefined.

Comment (leonprou) [reply alm/scripts/createSnapshots.js]: why also `blockLimit` is renamed to chunks after parseInt?
I think something like `blockSize` or `batchSize` suites more

**Review by leonprou — CHANGES_REQUESTED** [review]
Wrote some small fixes about names and etc..
What I do want:
In case the ALM doens't use snapshoting but searches live, as I understand we can't do batching over whole period. I suggest to batch from the current block to the start, and once we found the relevant event we can return it and halt. In such case in wortst case scenario it's gonna still be O(n), but for most of the searches it shoukd work alright

File: alm/.env.example (hunk 2 above)
Comment (t0mcr8se) [reply alm/.env.example]: how about we leave it empty if there's no block limit?

File: alm/src/utils/limitedEvents.ts (hunk 3 above)
Comment (t0mcr8se) [reply alm/src/utils/limitedEvents.ts:6]: hmmm, yes, once in js and once in ts, this one you're commenting on is a bit different than the one in the createSnapshots.js

but I can try to uniform them

File: alm/scripts/createSnapshots.js (hunk 4 above)
Comment (t0mcr8se) [reply alm/scripts/createSnapshots.js]: done
https://github.com/fuseio/tokenbridge/pull/6/commits/70960e6da2ccf089e75e1cd4dfb261eefe978318

File: alm/scripts/createSnapshots.js (hunk 5 above)
Comment (t0mcr8se) [reply alm/scripts/createSnapshots.js]: https://github.com/fuseio/tokenbridge/pull/6/commits/2f94988c3ccb35c167396518d714facb4551e936
is this better?

File: alm/.env.example (hunk 2 above)
Comment (t0mcr8se) [reply alm/.env.example]: https://github.com/fuseio/tokenbridge/pull/6/commits/2f94988c3ccb35c167396518d714facb4551e936

File: alm/.env.example (hunk 1 above)
Comment (t0mcr8se) [reply alm/.env.example]: https://github.com/fuseio/tokenbridge/pull/6/commits/b1d0174f9ff317c3c8ae676fa2c63dbfd4ed5535

**Pull request closed without merging**